_posts/2020-10-02-producer-tuning.md

```diff
@@ -0,0 +1,323 @@
+---
+layout: post
+title:  "Optimizing Kafka producers"
+date: 2020-10-02
+author: paul_mellor
+---
+
+You can fine tune Kafka producers using configuration properties to optimize the streaming of data to consumers.
+
+Get the tuning right, and even a small adjustment to your producer configuration can make a significant improvement to the way your producers operate.
+
+In this post we'll discuss typical tuning considerations for Kafka producers.
+
+<!--more-->
+
+### Optimizing Kafka producers
+
+Obviously, we want our producers to deliver data to Kafka topics as efficiently as possible.
+
+But what do we mean by this, and how do we quantify it?
+
+Do we base this on the number of messages sent over a set period of time? Or on how producers are set up to handle failure?
+
+Before starting your adventure in optimization, think about your destination. What are the results your are hoping to achieve?
+
+Think long enough about this, and you might find competing requirements. For example, by maximizing throughput you might also increase latency.
+
+Be prepared to make adjustments to your adjustments.
+
+### How is your producer performing?
+
+It's only when you have been monitoring the performance of your producers for some time that you can gauge how best to tune their performance.
+
+To begin with, you might start with a basic producer configuration in development as a benchmark. When you start to analyze producer metrics to see how the producers actually perform in typical production scenarios, you can make incremental changes and make comparisons until you hit the sweet spot.
```

Comment:
maybe mentioning the producer metrics we can put a reference to the producer monitoring documentation on the Kafka website? https://kafka.apache.org/documentation/#producer_sender_monitoring

Reply:
good idea. Added:

> If you want to read more about performance metrics for monitoring Kafka producers, see Kafka's [Producer Sender Metrics](https://kafka.apache.org/documentation/#producer_sender_monitoring).